### PR TITLE
Added property for user timezone [JENKINS-19887] [JENKINS-21143] [JEN…

### PR DESCRIPTION
**Pull request by chrissy25**
…KINS-1962]

Added the possibility that the users can select their own timezone to
display the timestamps correct for each timezone.

### CODE REVIEW

**Comment by olivergondza** [comment]
Related to: https://github.com/jenkinsci/jenkins/pull/580

File: core/src/main/java/hudson/timezone/DisplayTimezoneProperty.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
+/*
+ * The MIT License
+ * 
+ * Copyright (c) 2004-2010, Sun Microsystems, Inc., Tom Huybrechts
```
Comment (KostyaSha) [inline core/src/main/java/hudson/timezone/DisplayTimezoneProperty.java]: Really?

File: core/src/main/resources/hudson/timezone/DisplayTimezoneProperty/config.jelly [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<!--
+The MIT License
+
+Copyright (c) 2004-2011, Sun Microsystems, Inc., Tom Huybrechts, Alan Harder
```
Comment (KostyaSha) [inline core/src/main/resources/hudson/timezone/DisplayTimezoneProperty/config.jelly]: ?

**Comment by KostyaSha** [comment]
imho better rename variables with `TZ` instead long `Timezone`, `display` -> `show`

File: core/src/main/java/hudson/timezone/DisplayTimezoneProperty.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        this.displayTimezone = displayTimezone;
+    }
+
+    @Exported
+    public String getDisplayTimezone() {
+        return displayTimezone;
+    }
+
+    @Exported
+    public boolean getUseTimezone() {
```
Comment (KostyaSha) [inline core/src/main/java/hudson/timezone/DisplayTimezoneProperty.java]: If it boolean, then method should start from `is*`

**Comment by KostyaSha** [comment]
:-1: No javadocs, extra-spaces in coding-style, wrong license header(?), wrong method naming for primitive, variable naming under question

File: core/src/main/java/hudson/timezone/DisplayTimezoneProperty.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+            this.TIMEZONES = Arrays.asList( timeZones );
+        }
+
+        public String getDisplayName() {
+            return Messages.DisplayTimezoneProperty_DisplayName();
+        }
+
+        public UserProperty newInstance(User user) {
+            return new DisplayTimezoneProperty( false, TIMEZONES.get(0) ); //default setting is first time zone listed
```
Comment (jtnord) [inline core/src/main/java/hudson/timezone/DisplayTimezoneProperty.java]: default should the the system timezone - which means if you save this without making changes the behavior is consistent.

**Comment by oleg-nenashev** [comment]
@chrissy25 
Do you consider updating the pull-request according to the feedback?

**Comment by chrissy25** [comment]
Yes I will, didn't had time until now. But will do it this week :)

**Comment by oleg-nenashev** [comment]
@chrissy25 
Thanks! this feature could be really useful. Consider integrating it with #580 , which provides a generic engine for it.

**Comment by chrissy25** [comment]
Added the changes.
Main question would be now, do we want to have a solution by a plugin or directly in the core?
I am open minded in this direction :)

File: core/src/main/resources/hudson/model/AbstractBuild/index.jelly [hunk 5]
```diff
@@ -25,6 +25,13 @@ THE SOFTWARE.
 
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+  <j:if test="${!h.isAnonymous()}">
+    <j:invokeStatic var="userHasDisplayTimeZone" className="hudson.timezone.DisplayTimezoneProperty" method="isCurrentUserUseDisplayTimezone" />
+    <j:if test="${userHasDisplayTimeZone}">
+      <j:invokeStatic var="userDisplayTimeZone" className="hudson.timezone.DisplayTimezoneProperty" method="getCurrentUserDisplayTimezone" />
+      <i:setTimeZone value="${userDisplayTimeZone}"/>
+    </j:if>
+  </j:if>
```
Comment (olivergondza) [inline core/src/main/resources/hudson/model/AbstractBuild/index.jelly]: Can this be in `layout.jelly` so we do not have to put it in all the views (or even worse, present different dates in different timezones causing confusion)?

**Comment by oleg-nenashev** [comment]
> Main question would be now, do we want to have a solution by a plugin or directly in the core?
I am open minded in this direction :)

I'm OK to integrate it into the core, but a common trend is to keep features in plugins if it is possible

**Comment by oleg-nenashev** [comment]
Uhm, merge conflict :(

BTW :+1:  from me when it gets fixed

**Comment by daniel-beck** [comment]
@chrissy25 Are you still seeking to have this PR integrated? If so, please let us know and rebase it into recent master, resolving the merge conflict.

**Comment by batmat** [comment]
I'm closing this PR as per the last comment without answer and the label added back then by Daniel.

@chrissy25 thanks a lot for the work. If you find the time and desire to get back to this, feel obviously warmly welcome to do so. 

Thanks!

**Pull request closed without merging**